Panic on send certificates if missing contact

There is a bug in the sending certificates code that caused a panic, possibly because there was no billing contact. ﻿﻿
﻿This also means that Rotational's status was not correctly updated. ﻿
﻿
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x8fc0a6]

goroutine 654 [running]:
github.com/trisacrypto/directory/pkg/gds/models/v1.GetContactVerification(0x0, 0xc00044c4b0, 0x0, 0x0, 0xc00044f380, 0x11)
	/srv/build/pkg/gds/models/v1/models.go:38 +0x26
github.com/trisacrypto/directory/pkg/gds/emails.(*EmailManager).SendDeliverCertificates(0xc00032e460, 0xc0000cda40, 0xc00003aa68, 0x11, 0xc00044c0f0, 0x0)
	/srv/build/pkg/gds/emails/client.go:185 +0x22b
github.com/trisacrypto/directory/pkg/gds.(*Server).downloadCertificateRequest(0xc00042d040, 0xc0000f8f00)
	/srv/build/pkg/gds/certs.go:329 +0xbd6
created by github.com/trisacrypto/directory/pkg/gds.(*Server).checkCertificateRequest
	/srv/build/pkg/gds/certs.go:226 +0x5f7
```

--- a/go.mod
+++ b/go.mod
@@ -6,6 +6,7 @@ require (
 	cloud.google.com/go v0.82.0
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/dgrijalva/jwt-go v3.2.0+incompatible
+	github.com/ghodss/yaml v1.0.0
 	github.com/golang/protobuf v1.5.2
 	github.com/google/uuid v1.2.0
 	github.com/joho/godotenv v1.3.0

--- a/pkg/gds/certs.go
+++ b/pkg/gds/certs.go
@@ -327,7 +327,9 @@ func (s *Server) downloadCertificateRequest(r *models.CertificateRequest) {
 
 	// Email the certificates to the technical contacts
 	if err = s.email.SendDeliverCertificates(vasp, path); err != nil {
+		// If there is an error delivering emails, return here so we don't mark as completed
 		log.Error().Err(err).Msg("could not deliver certificates to technical contact")
+		return
 	}
 
 	// Mark certificate request as complete.

--- a/pkg/gds/emails/client.go
+++ b/pkg/gds/emails/client.go
@@ -2,17 +2,19 @@ package emails
 
 import (
 	"encoding/hex"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"net/mail"
 
+	"github.com/ghodss/yaml"
+	"github.com/rs/zerolog/log"
 	"github.com/sendgrid/rest"
 	"github.com/sendgrid/sendgrid-go"
 	sgmail "github.com/sendgrid/sendgrid-go/helpers/mail"
 	"github.com/trisacrypto/directory/pkg/gds/config"
 	"github.com/trisacrypto/directory/pkg/gds/models/v1"
 	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+	"google.golang.org/protobuf/encoding/protojson"
 )
 
 // New email manager with the specified configuration.
@@ -64,7 +66,12 @@ func (m *EmailManager) SendVerifyContacts(vasp *pb.VASP) (err error) {
 		vasp.Contacts.Billing, vasp.Contacts.Legal,
 	}
 
-	for _, contact := range contacts {
+	// Attempt at least one delivery, don't give up just because one email failed
+	// Track how many emails and errors occurred during delivery.
+	var nDelivered, nErrors uint8
+
+	for idx, contact := range contacts {
+		// Skip any null contacts or contacts without email addresses
 		if contact == nil || contact.Email == "" {
 			continue
 		}
@@ -74,6 +81,7 @@ func (m *EmailManager) SendVerifyContacts(vasp *pb.VASP) (err error) {
 			VID:  vasp.Id,
 		}
 		if ctx.Token, _, err = models.GetContactVerification(contact); err != nil {
+			// If we can't get the verification token, this is a fatal error
 			return err
 		}
 
@@ -83,23 +91,47 @@ func (m *EmailManager) SendVerifyContacts(vasp *pb.VASP) (err error) {
 			ctx,
 		)
 		if err != nil {
-			return err
+			log.Error().Err(err).Str("vasp", vasp.Id).Int("contact", idx).Msg("could not create verify contact email")
+			nErrors++
+			continue
 		}
 
 		if err = m.Send(msg); err != nil {
-			return err
+			log.Error().Err(err).Str("vasp", vasp.Id).Int("contact", idx).Msg("could not send verify contact email")
+			nErrors++
+			continue
 		}
+
+		nDelivered++
 	}
 
+	// Return an error if no emails were delivered
+	if nDelivered == 0 {
+		return fmt.Errorf("no verify contact emails were successfully sent (%d errors)", nErrors)
+	}
 	return nil
 }
 
 // SendReviewRequest is a shortcut for iComply verification in which we simply send
 // an email to the TRISA admins and have them manually verify registrations.
 func (m *EmailManager) SendReviewRequest(vasp *pb.VASP) (err error) {
+	jsonpb := protojson.MarshalOptions{
+		Multiline:       false,
+		AllowPartial:    true,
+		UseProtoNames:   true,
+		UseEnumNumbers:  false,
+		EmitUnpopulated: true,
+	}
+
 	var data []byte
-	if data, err = json.MarshalIndent(vasp, "", "  "); err != nil {
+	if data, err = jsonpb.Marshal(vasp); err != nil {
 		return err
+	}
+
+	// Convert JSON to YAML to make it more human readable
+	// If the conversion fails, then the JSON data will be kept
+	if yamlData, err := yaml.JSONToYAML(data); err == nil {
+		data = yamlData
 	}
 
 	ctx := ReviewRequestData{
@@ -139,13 +171,23 @@ func (m *EmailManager) SendRejectRegistration(vasp *pb.VASP, reason string) (err
 		vasp.Contacts.Billing, vasp.Contacts.Legal,
 	}
 
-	for _, contact := range contacts {
+	// Attempt at least one delivery, don't give up just because one email failed
+	// Track how many emails and errors occurred during delivery.
+	var nDelivered, nErrors uint8
+
+	for idx, contact := range contacts {
+		// Skip any contacts that we can't send emails to
+		if contact == nil || contact.Email == "" {
+			continue
+		}
+
 		var verified bool
 		if _, verified, err = models.GetContactVerification(contact); err != nil {
+			// If we can't get the verification, this is a fatal error
 			return err
 		}
 
-		if contact != nil && verified {
+		if verified {
 			ctx.Name = contact.Name
 			msg, err := RejectRegistrationEmail(
 				m.serviceEmail.Name, m.serviceEmail.Address,
@@ -153,15 +195,25 @@ func (m *EmailManager) SendRejectRegistration(vasp *pb.VASP, reason string) (err
 				ctx,
 			)
 			if err != nil {
-				return err
+				log.Error().Err(err).Str("vasp", vasp.Id).Int("contact", idx).Msg("could not create reject registration email")
+				nErrors++
+				continue
 			}
 
 			if err = m.Send(msg); err != nil {
-				return err
+				log.Error().Err(err).Str("vasp", vasp.Id).Int("contact", idx).Msg("could not send reject registration email")
+				nErrors++
+				continue
 			}
+
+			nDelivered++
 		}
 	}
 
+	// Return an error if no emails were delivered
+	if nDelivered == 0 {
+		return fmt.Errorf("no registration rejection emails were successfully sent (%d errors)", nErrors)
+	}
 	return nil
 }
 
@@ -180,12 +232,23 @@ func (m *EmailManager) SendDeliverCertificates(vasp *pb.VASP, path string) (err 
 		vasp.Contacts.Billing, vasp.Contacts.Legal,
 	}
 
-	for _, contact := range contacts {
+	// Attempt at least one delivery, don't give up just because one email failed
+	// Track how many emails and errors occurred during delivery.
+	var nDelivered, nErrors uint8
+
+	for idx, contact := range contacts {
+		// Skip any null contacts or contacts without email addresses
+		if contact == nil || contact.Email == "" {
+			continue
+		}
+
 		var verified bool
 		if _, verified, err = models.GetContactVerification(contact); err != nil {
+			// If we can't get the verification this is a fatal error
 			return err
 		}
-		if contact != nil && verified {
+
+		if verified {
 			ctx.Name = contact.Name
 			msg, err := DeliverCertsEmail(
 				m.serviceEmail.Name, m.serviceEmail.Address,
@@ -194,14 +257,24 @@ func (m *EmailManager) SendDeliverCertificates(vasp *pb.VASP, path string) (err 
 			)
 
 			if err != nil {
-				return err
+				log.Error().Err(err).Str("vasp", vasp.Id).Int("contact", idx).Msg("could not create deliver certs email")
+				nErrors++
+				continue
 			}
 
 			if err = m.Send(msg); err != nil {
-				return err
+				log.Error().Err(err).Str("vasp", vasp.Id).Int("contact", idx).Msg("could not send deliver certs email")
+				nErrors++
+				continue
 			}
+
+			nDelivered++
 		}
 	}
 
+	// Return an error if no emails were delivered
+	if nDelivered == 0 {
+		return fmt.Errorf("no certificate delivery emails were successfully sent (%d errors)", nErrors)
+	}
 	return nil
 }

--- a/pkg/gds/emails/client_test.go
+++ b/pkg/gds/emails/client_test.go
@@ -1,0 +1,63 @@
+package emails_test
+
+import (
+	"net/mail"
+	"os"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/joho/godotenv"
+	"github.com/kelseyhightower/envconfig"
+	"github.com/stretchr/testify/require"
+	"github.com/trisacrypto/directory/pkg/gds/config"
+	"github.com/trisacrypto/directory/pkg/gds/emails"
+	"github.com/trisacrypto/directory/pkg/gds/models/v1"
+	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+)
+
+func TestClientSendEmails(t *testing.T) {
+	// NOTE: if you place a .env file in this directory alongside the test file, it
+	// will be read, making it simpler to run tests and set environment variables.
+	godotenv.Load()
+
+	if os.Getenv("GDS_TEST_SENDING_CLIENT_EMAILS") == "" {
+		t.Skip("skip client send emails test")
+	}
+
+	// This test uses the environment to send rendered emails with context specific
+	// emails - this is to test the rendering of the emails with data only; it does not
+	// go through any of the server workflow for generating tokens, etc.
+	var conf config.EmailConfig
+	err := envconfig.Process("gds", &conf)
+	require.NoError(t, err)
+
+	// This test sends emails from the serviceEmail using SendGrid to the adminsEmail
+	email, err := emails.New(conf)
+	require.NoError(t, err)
+
+	receipient, err := mail.ParseAddress(conf.AdminEmail)
+	require.NoError(t, err)
+
+	vasp := &pb.VASP{
+		Id:            uuid.NewString(),
+		CommonName:    "test.example.com",
+		TrisaEndpoint: "test.example.com:443",
+		Contacts: &pb.Contacts{
+			Technical: &pb.Contact{
+				Name:  receipient.Name,
+				Email: receipient.Address,
+			},
+		},
+		IdentityCertificate: &pb.Certificate{
+			SerialNumber: []byte("notarealcertificate"),
+		},
+	}
+
+	err = models.SetContactVerification(vasp.Contacts.Technical, "", true)
+	require.NoError(t, err)
+
+	require.NoError(t, email.SendVerifyContacts(vasp))
+	require.NoError(t, email.SendReviewRequest(vasp))
+	require.NoError(t, email.SendRejectRegistration(vasp, "this is a test rejection from the test runner"))
+	require.NoError(t, email.SendDeliverCertificates(vasp, "testdata/foo.zip"))
+}

--- a/pkg/gds/emails/emails_test.go
+++ b/pkg/gds/emails/emails_test.go
@@ -74,11 +74,13 @@ func TestVerifyContactURL(t *testing.T) {
 }
 
 func TestSendEmails(t *testing.T) {
+	// NOTE: if you place a .env file in this directory alongside the test file, it
+	// will be read, making it simpler to run tests and set environment variables.
+	godotenv.Load()
+
 	if os.Getenv("GDS_TEST_SENDING_EMAILS") == "" {
 		t.Skip("skip generate and send emails test")
 	}
-
-	godotenv.Load()
 
 	// This test uses the environment to send rendered emails with context specific
 	// emails - this is to test the rendering of the emails with data only; it does not

--- a/pkg/gds/emails/templates/review_request.html
+++ b/pkg/gds/emails/templates/review_request.html
@@ -3,7 +3,7 @@
 <p>We have received a new TRISA Global Directory registration request from a VASP that
 needs to be reviewed. The requestor has verified their email address(es) and received a
 PKCS12 password to decrypt a certificate which will be generated only if you approve this
-request. The request JSON is:</p>
+request. The request YAML is:</p>
 
 <pre>{{ .Request }}</pre>
 

--- a/pkg/gds/emails/templates/review_request.txt
+++ b/pkg/gds/emails/templates/review_request.txt
@@ -3,7 +3,7 @@ Hello TRISA Admins,
 We have received a new TRISA Global Directory registration request from a VASP that
 needs to be reviewed. The requestor has verified their email address(es) and received a
 PKCS12 password to decrypt a certificate which will be generated only if you approve this
-request. The request JSON is:
+request. The request YAML is:
 
 {{ .Request }}
 

